### PR TITLE
Restore null behavior

### PR DESCRIPTION
**Pull request by electrum**

### CODE REVIEW

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java [hunk 1]
```diff
@@ -1275,8 +1275,10 @@ public void testLeast()
     public void testGreatestWithNaN()
     {
         assertFunction("greatest(1.5E0, 0.0E0 / 0.0E0)", DOUBLE, Double.NaN);
+        assertFunction("greatest(1.5E0, 0.0E0 / 0.0E0, 2.7E0)", DOUBLE, Double.NaN);
         assertFunction("greatest(1.5E0, REAL '0.0' / REAL '0.0')", DOUBLE, Double.NaN);
-        assertFunction("greatest(null, REAL '0.0' / REAL '0.0')", REAL, Float.NaN);
+        assertFunction("greatest(1.5E0, REAL '0.0' / REAL '0.0', 2.7E0)", DOUBLE, Double.NaN);
+        assertFunction("greatest(null, REAL '0.0' / REAL '0.0')", REAL, null);
```
Comment (martint) [inline presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java:1281]: Unrelated to this commit

**Review by martint — APPROVED** [review]

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java [hunk 2]
```diff
@@ -1142,42 +1142,42 @@ public void testGreatest()
         assertFunction("greatest(TINYINT'-1', TINYINT'-2')", TINYINT, (byte) -1);
         assertFunction("greatest(TINYINT'5', TINYINT'4', TINYINT'3', TINYINT'2', TINYINT'1', TINYINT'2', TINYINT'3', TINYINT'4', TINYINT'1', TINYINT'5')", TINYINT, (byte) 5);
         assertFunction("greatest(TINYINT'-1')", TINYINT, (byte) -1);
-        assertFunction("greatest(TINYINT'5', TINYINT'4', CAST(NULL AS TINYINT), TINYINT'3')", TINYINT, (byte) 5);
+        assertFunction("greatest(TINYINT'5', TINYINT'4', CAST(NULL AS TINYINT), TINYINT'3')", TINYINT, null);
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java:1145]: You do not even need to add a new test, nice.

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java [hunk 3]
```diff
@@ -1142,42 +1142,42 @@ public void testGreatest()
...
 
         // double
         assertFunction("greatest(1.5E0, 2.3E0)", DOUBLE, 2.3);
         assertFunction("greatest(-1.5E0, -2.3E0)", DOUBLE, -1.5);
         assertFunction("greatest(-1.5E0, -2.3E0, -5/3)", DOUBLE, -1.0);
         assertFunction("greatest(1.5E0, -1.0E0 / 0.0E0, 1.0E0 / 0.0E0)", DOUBLE, Double.POSITIVE_INFINITY);
-        assertFunction("greatest(5, 4, CAST(NULL as DOUBLE), 3)", DOUBLE, 5.0);
+        assertFunction("greatest(5, 4, CAST(NULL as DOUBLE), 3)", DOUBLE, null);
 
         // float
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java:1175]: `// real`

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java [hunk 4]
```diff
@@ -1142,42 +1142,42 @@ public void testGreatest()
...
         assertFunction("greatest(REAL '1.5', 2.3E0)", DOUBLE, 2.3);
         assertFunction("greatest(REAL '-1.5', -2.3E0)", DOUBLE, (double) -1.5f);
         assertFunction("greatest(-1.5E0, REAL '-2.3', -5/3)", DOUBLE, -1.0);
         assertFunction("greatest(REAL '1.5', REAL '-1.0' / 0.0E0, 1.0E0 / REAL '0.0')", DOUBLE, (double) (1.0f / 0.0f));
-        assertFunction("greatest(5, REAL '4', CAST(NULL as DOUBLE), 3)", DOUBLE, 5.0);
+        assertFunction("greatest(5, REAL '4', CAST(NULL as DOUBLE), 3)", DOUBLE, null);
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java:1180]: This is a test case for DOUBLE.
Make it into a test case for REAL:

```suggestion
        assertFunction("greatest(5, REAL '4', CAST(NULL AS real), 3)", REAL, null);
```

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java [hunk 5]
```diff
@@ -1142,42 +1142,42 @@ public void testGreatest()
...
-        assertFunction("greatest(5000000000, 4000000000, CAST(NULL as BIGINT), 3000000000)", BIGINT, 5000000000L);
+        assertFunction("greatest(5000000000, 4000000000, CAST(NULL as BIGINT), 3000000000)", BIGINT, null);
 
         // double
         assertFunction("greatest(1.5E0, 2.3E0)", DOUBLE, 2.3);
         assertFunction("greatest(-1.5E0, -2.3E0)", DOUBLE, -1.5);
         assertFunction("greatest(-1.5E0, -2.3E0, -5/3)", DOUBLE, -1.0);
         assertFunction("greatest(1.5E0, -1.0E0 / 0.0E0, 1.0E0 / 0.0E0)", DOUBLE, Double.POSITIVE_INFINITY);
-        assertFunction("greatest(5, 4, CAST(NULL as DOUBLE), 3)", DOUBLE, 5.0);
+        assertFunction("greatest(5, 4, CAST(NULL as DOUBLE), 3)", DOUBLE, null);
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java:1173]: Add a greatest and least test cases for DOUBLE and REAL with 1,NaN,3 and 1,NULL,NaN,4

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java [hunk 6]
```diff
@@ -1190,8 +1190,8 @@ public void testGreatest()
         assertFunction("greatest(1.0E0, 2)", DOUBLE, 2.0);
         assertFunction("greatest(1, 2.0E0)", DOUBLE, 2.0);
         assertFunction("greatest(1.0E0, 2)", DOUBLE, 2.0);
-        assertFunction("greatest(5.0E0, 4, CAST(NULL as DOUBLE), 3)", DOUBLE, 5.0);
-        assertFunction("greatest(5.0E0, 4, CAST(NULL as BIGINT), 3)", DOUBLE, 5.0);
+        assertFunction("greatest(5.0E0, 4, CAST(NULL as DOUBLE), 3)", DOUBLE, null);
+        assertFunction("greatest(5.0E0, 4, CAST(NULL as BIGINT), 3)", DOUBLE, null);
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java:1194]: Duplicate test case?

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java [hunk 7]
```diff
@@ -1214,42 +1214,42 @@ public void testLeast()
...
 
         // double
         assertFunction("least(1.5E0, 2.3E0)", DOUBLE, 1.5);
         assertFunction("least(-1.5E0, -2.3E0)", DOUBLE, -2.3);
         assertFunction("least(-1.5E0, -2.3E0, -5/3)", DOUBLE, -2.3);
         assertFunction("least(1.5E0, -1.0E0 / 0.0E0, 1.0E0 / 0.0E0)", DOUBLE, Double.NEGATIVE_INFINITY);
-        assertFunction("least(5, 4, CAST(NULL as DOUBLE), 3)", DOUBLE, 3.0);
+        assertFunction("least(5, 4, CAST(NULL as DOUBLE), 3)", DOUBLE, null);
 
         // float
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java:1247]: `// real`

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java [hunk 8]
```diff
@@ -1214,42 +1214,42 @@ public void testLeast()
...
         assertFunction("least(REAL '1.5', 2.3E0)", DOUBLE, (double) 1.5f);
         assertFunction("least(REAL '-1.5', -2.3E0)", DOUBLE, -2.3);
         assertFunction("least(-2.3E0, REAL '-0.4', -5/3)", DOUBLE, -2.3);
         assertFunction("least(1.5E0, REAL '-1.0' / 0.0E0, 1.0E0 / 0.0E0)", DOUBLE, (double) (-1.0f / 0.0f));
-        assertFunction("least(REAL '5', 4, CAST(NULL as DOUBLE), 3)", DOUBLE, 3.0);
+        assertFunction("least(REAL '5', 4, CAST(NULL as DOUBLE), 3)", DOUBLE, null);
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java:1252]: This is a test case for DOUBLE.
Make it into a test case for REAL:

```suggestion
        assertFunction("least(REAL '5', 4, CAST(NULL AS real), 3)", REAL, null);
```

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java (hunk 4 above)
Comment (findepi) [reply presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java:1180]: Actually the cases above are also for DOUBLE, not REAL

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java (hunk 8 above)
Comment (findepi) [reply presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java:1252]: Actually the cases above are also for DOUBLE, not REAL

File: presto-main/src/test/java/io/prestosql/type/TestArrayOperators.java [hunk 9]
```diff
@@ -597,13 +597,13 @@ public void testArrayMin()
         assertFunction("ARRAY_MIN(ARRAY [NaN()])", DOUBLE, NaN);
         assertFunction("ARRAY_MIN(ARRAY [NULL, NULL, NULL])", UNKNOWN, null);
         assertFunction("ARRAY_MIN(ARRAY [NaN(), NaN(), NaN()])", DOUBLE, NaN);
-        assertFunction("ARRAY_MIN(ARRAY [NULL, 2, 3])", INTEGER, 2);
+        assertFunction("ARRAY_MIN(ARRAY [NULL, 2, 3])", INTEGER, null);
         assertFunction("ARRAY_MIN(ARRAY [NaN(), 2, 3])", DOUBLE, 2.0);
-        assertFunction("ARRAY_MIN(ARRAY [NULL, NaN(), 1])", DOUBLE, 1.0);
-        assertFunction("ARRAY_MIN(ARRAY [NaN(), NULL, 3.0])", DOUBLE, 3.0);
-        assertFunction("ARRAY_MIN(ARRAY [1.0E0, NULL, 3])", DOUBLE, 1.0E0);
+        assertFunction("ARRAY_MIN(ARRAY [NULL, NaN(), 1])", DOUBLE, null);
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/type/TestArrayOperators.java:602]: Before c95e937517f773ba625cfde38aac7ed11e6f8493 the expected values here were different -- here and in some other places NaN was expected. See

https://github.com/prestosql/presto/blob/f04fa5ca8de1015629a94693740dfbc641e55784/presto-main/src/test/java/io/prestosql/type/TestArrayOperators.java#L608-L612

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java (hunk 6 above)
Comment (electrum) [reply presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java:1194]: Duplicate with what? They are different input types

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java (hunk 1 above)
Comment (electrum) [reply presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java:1281]: It's testing that the loop condition handling is correct when there are more than two values.

File: presto-main/src/test/java/io/prestosql/type/TestArrayOperators.java (hunk 9 above)
Comment (electrum) [reply presto-main/src/test/java/io/prestosql/type/TestArrayOperators.java:602]: You're right, though that seems like a different issue, and a bug. It's not clear why NaN would be returned over NULL, but non-NaN values would return NULL.

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java (hunk 2 above)
Comment (electrum) [reply presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java:1145]: I'm not sure what this comment means.

File: presto-main/src/test/java/io/prestosql/type/TestArrayOperators.java (hunk 9 above)
Comment (electrum) [reply presto-main/src/test/java/io/prestosql/type/TestArrayOperators.java:602]: According to the documentation, it is a bug:

> Like most other functions in Presto, they return null if any argument is null.

Comment (findepi) [reply presto-main/src/test/java/io/prestosql/type/TestArrayOperators.java:602]: I think the logic can be "return NaN when any of the compared values is NaN" and "NULL means unknown, so return unknown result when any of the compared values is unknown but, unless the result will be NaN regardless of the unknown value"

Comment (findepi) [reply presto-main/src/test/java/io/prestosql/type/TestArrayOperators.java:602]: Anyway, even if want to change the behavior, I'd strongly prefer to restore pre-344 behavior to clear "release blocker" and then fix whatever remains to be fixed on top of that.

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java (hunk 6 above)
Comment (findepi) [reply presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java:1194]: The two lines
```
assertFunction("greatest(5.0E0, 4, CAST(NULL as DOUBLE), 3)", DOUBLE, null);
assertFunction("greatest(5.0E0, 4, CAST(NULL as BIGINT), 3)", DOUBLE, null);
```

are probably equivalent from `greatest` function perspective (with and without a coercion, but does that matter?)

**Review by dain — APPROVED** [review]
Looks good to me

File: presto-main/src/test/java/io/prestosql/type/TestArrayOperators.java (hunk 9 above)
Comment (electrum) [reply presto-main/src/test/java/io/prestosql/type/TestArrayOperators.java:602]: Actually, that quote is from greatest/least. The only documentation for array_min is

> Returns the minimum value of input array.

**Comment by electrum** [comment]
I'm going to merge this PR as-is since it's about restoring null behavior, and the comments are about deficiencies in existing tests, or in restoring the NaN behavior for the array functions. We can follow up on those separately.

**Comment by findepi** [comment]
Can you please create issue(s) for the outstanding work?

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java (hunk 4 above)
Comment (findepi) [reply presto-main/src/test/java/io/prestosql/operator/scalar/TestMathFunctions.java:1180]: This remained unanswered, so we don't have a test for NULL with REAL type just yet.
